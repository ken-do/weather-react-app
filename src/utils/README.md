# Utils

Reusable helper functions, constants and higher-order-components created to be used flexibly throughout the app. These files include:

### api

A http client based on axios. Some default options were added for convenience and proper errors handling were implemented.

### constants

App constants are being defined in this file.

### formatters

Helper functions used to convert values from one format into another, for example, we use `dateToFriendlyDate` to convert a ISO formatted date string `2021-02-07` into a more readable and user-friendly string `Sun Feb 07 2021`.

### normalizers

Functions written to make sure predefined coding/naming conventions are met. For example, we requires that variable and object keys should be in `camelCase`, but an external services might return data that uses `snake_case` for it attribute keys. In this case, we can use a normalizer to convert the attributes keys into `camelCase` for the sake of consistency.

### router

Helper functions that enables quick interaction with the app router state. An example is `useQuery`, which allows us to get the current URL query params more quickly, since `react-router-dom` does not provide a quick way achieve this.

### ssr

Helper functions that assists the server-side rendering process.

### store

Typed versions of common `react-redux` hooks. We export these hooks for usage accross the app, so that we don't have to repeatedly type the original functions every time we call `useDispatch` and `useSelector`.

### testUtils

Defines a custom render that uses react-testing-library `render` underlyingly and expected to replace testing-library `render` in unit tests.
