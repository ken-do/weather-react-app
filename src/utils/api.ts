import axios, { AxiosRequestConfig } from 'axios'

const defaultConfig: AxiosRequestConfig = {
    baseURL: process.env.REACT_APP_BASE_API_URL,
    timeout: 2500,
    headers: {
        Accepts: 'application/json',
        'Content-Security-Policy':
            "default-src 'self' metaweather.com *.metaweather.com",
    },
}

interface HttpErrorMessages {
    [key: number]: string
}

export const httpErrorMessages: HttpErrorMessages = {
    400: 'Bad Request',
    401: 'Unauthorized',
    402: 'Payment Required',
    403: 'Forbidden',
    404: 'Not Found',
    405: 'Method Not Allowed',
    406: 'Not Acceptable',
    407: 'Proxy Authentication Required',
    408: 'Request Timeout',
    409: 'Conflict',
    410: 'Gone',
    411: 'Length Required',
    412: 'Precondition Failed',
    413: 'Request Entity Too Large',
    414: 'Request-URI Too Long',
    415: 'Unsupported Media Type',
    416: 'Requested Range Not Satisfiable',
    417: 'Expectation Failed',
    418: "I'm a teapot (RFC 2324)",
    420: 'Enhance Your Calm (Twitter)',
    422: 'Unprocessable Entity (WebDAV)',
    423: 'Locked (WebDAV)',
    424: 'Failed Dependency (WebDAV)',
    425: 'Reserved for WebDAV',
    426: 'Upgrade Required',
    428: 'Precondition Required',
    429: 'Too Many Requests',
    431: 'Request Header Fields Too Large',
    444: 'No Response (Nginx)',
    449: 'Retry With (Microsoft)',
    450: 'Blocked by Windows Parental Controls (Microsoft)',
    451: 'Unavailable For Legal Reasons',
    499: 'Client Closed Request (Nginx)',
    500: 'Internal Server Error',
    501: 'Not Implemented',
    502: 'Bad Gateway',
    503: 'Service Unavailable',
    504: 'Gateway Timeout',
    505: 'HTTP Version Not Supported',
    506: 'Variant Also Negotiates (Experimental)',
    507: 'Insufficient Storage (WebDAV)',
    508: 'Loop Detected (WebDAV)',
    509: 'Bandwidth Limit Exceeded (Apache)',
    510: 'Not Extended',
    511: 'Network Authentication Required',
}

export const endpoints = {
    suggestion: '/location/search',
    location: '/location',
}

const api = axios.create(defaultConfig)

api.interceptors.response.use(
    (response) => response,
    (error) => {
        const standardizedError = new Error()
        // The request was made and the server responded with a status code
        if (error.response) {
            standardizedError.message = httpErrorMessages[error.response.status]
        } else if (error.message) {
            // Wrong request setting
            standardizedError.message = error.message
            // The request was made but no response was received
        } else {
            standardizedError.message = 'Server did not respond.'
        }
        return Promise.reject(standardizedError)
    }
)

export default api
