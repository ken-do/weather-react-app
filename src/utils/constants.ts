// eslint-disable-next-line import/prefer-default-export
export const DEBOUNCE_WAIT_MS = 500
export const PATH = {
    home: '/',
    location: '/location',
    weather: '/weather',
}
export const MAX_SUGGESTIONS = 5
export const MAX_FORECASTS = 5
export const STATIC_IMG_ROOT = `${process.env.REACT_APP_API_ORIGIN}/static/img/weather/`
export const APP_NAME = 'Weather React App'
export const LOCATION_SEARCH_PLACEHOLDER = 'Enter a location...'
export const EVENT_KEYS = {
    enter: 'Enter',
}
