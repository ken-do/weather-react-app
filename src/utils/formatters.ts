export const dateToFriendlyDate = (date: string) =>
    new Date(date).toDateString()

export const UTCStringToHours = (time: string) =>
    new Date(time).toLocaleString('en-US', {
        hour: 'numeric',
        minute: 'numeric',
        hour12: true,
    })
