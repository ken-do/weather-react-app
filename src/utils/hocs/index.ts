/* eslint-disable import/prefer-default-export */
export { default as withSuggestions } from './withSuggestions'
