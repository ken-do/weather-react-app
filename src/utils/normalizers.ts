/* eslint-disable import/prefer-default-export */
import _camelCase from 'lodash/camelCase'
import _toPairs from 'lodash/toPairs'
import _fromPairs from 'lodash/fromPairs'

export const normalizeObjectKeys = <T extends Record<string, unknown>, R>(
    obj: T
): R => {
    const pairsFromObj = _toPairs(obj)
    const newPairs = pairsFromObj.map(([key, value]) => [
        _camelCase(key),
        value,
    ])
    return _fromPairs(newPairs) as R
}
