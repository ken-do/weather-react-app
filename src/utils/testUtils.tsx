import React, { PropsWithChildren, ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import store from 'src/store'
import { BrowserRouter as Router } from 'react-router-dom'
import { Provider } from 'react-redux'

export const AllTheProviders = ({ children }: PropsWithChildren<unknown>) => (
    <Provider store={store}>
        <Router>{children}</Router>
    </Provider>
)

const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, 'queries'>
) => {
    return render(ui, { wrapper: AllTheProviders, ...options })
}

export * from '@testing-library/react'

export { customRender as render }
