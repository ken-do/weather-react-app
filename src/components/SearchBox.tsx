import React, { memo, useState, useCallback } from 'react'
import Form from 'react-bootstrap/Form'
import FormControl from 'react-bootstrap/FormControl'
import InputGroup from 'react-bootstrap/InputGroup'
import Button from 'react-bootstrap/Button'
import Spinner from 'react-bootstrap/Spinner'
import { withSuggestions } from 'utils/hocs'
import { LOCATION_SEARCH_PLACEHOLDER, EVENT_KEYS } from 'utils/constants'
import styles from './SearchBox.module.scss'

export interface SearchBoxProps {
    afterInputChanged?: (value: string) => void
    afterButtonClicked?: (value: string) => void
    size?: 'sm' | 'lg' | undefined
    autoFocus?: boolean
    isLoading?: boolean
}

export const SearchBox: React.FC<SearchBoxProps> = ({
    afterInputChanged,
    afterButtonClicked,
    size,
    autoFocus,
    isLoading,
}: SearchBoxProps) => {
    const [value, setValue] = useState('')

    const handleInputChange = useCallback(
        (event) => {
            setValue(event.target.value)
            if (afterInputChanged) {
                afterInputChanged(event.target.value)
            }
        },
        [afterInputChanged]
    )

    const handleButtonClick = useCallback(() => {
        if (afterButtonClicked) {
            afterButtonClicked(value)
        }
        setValue('')
    }, [afterButtonClicked, value])

    const handleEnterKey = useCallback(
        (event) => {
            if (event.key === EVENT_KEYS.enter) {
                event.preventDefault()
                handleButtonClick()
            }
        },
        [handleButtonClick]
    )

    return (
        <Form inline className="flex-nowrap">
            <InputGroup size={size} className={styles.inputGroup}>
                <FormControl
                    type="text"
                    placeholder={LOCATION_SEARCH_PLACEHOLDER}
                    className="border-right-0 rounded-0"
                    value={value}
                    onChange={handleInputChange}
                    onKeyDown={handleEnterKey}
                    autoFocus={autoFocus}
                />
                <Button variant="success rounded-0" onClick={handleButtonClick}>
                    {isLoading ? (
                        <>
                            <Spinner
                                as="span"
                                animation="grow"
                                size="sm"
                                role="status"
                                aria-hidden="true"
                                className="mr-1"
                            />
                            Loading...
                        </>
                    ) : (
                        'Search'
                    )}
                </Button>
            </InputGroup>
        </Form>
    )
}

SearchBox.defaultProps = {
    afterInputChanged: () => {},
    afterButtonClicked: () => {},
    size: undefined,
    autoFocus: false,
    isLoading: false,
}

export const SearchBoxWithSuggestions = withSuggestions(SearchBox)

export default memo(SearchBox)
