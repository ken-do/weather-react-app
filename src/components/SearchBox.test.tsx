import { render, fireEvent } from 'src/utils/testUtils'
import React from 'react'
import { EVENT_KEYS } from 'src/utils/constants'
import { SearchBox } from './SearchBox'

describe('SearchBox', () => {
    test('should render an input field and a button', () => {
        const { getByRole } = render(<SearchBox />)
        const button = getByRole('button')
        const input = getByRole('textbox')
        expect(button).toBeInTheDocument()
        expect(input).toBeInTheDocument()
    })

    test('should focus on the input element if autoFocus prop is to set true', () => {
        const { getByRole } = render(<SearchBox autoFocus />)
        const input = getByRole('textbox')
        expect(input).toHaveFocus()
    })

    test('should enlarge when passing in a lg size prop', () => {
        const { asFragment, rerender } = render(<SearchBox />)
        const firstRender = asFragment()
        rerender(<SearchBox size="lg" />)
        expect(firstRender).toMatchDiffSnapshot(asFragment())
    })

    it('should update the button lable based on isLoading value', () => {
        const { asFragment, rerender, queryByText } = render(<SearchBox />)
        const firstRender = asFragment()
        const button = queryByText(/search/i)
        expect(button).toBeVisible()
        rerender(<SearchBox isLoading />)
        const loadingButton = queryByText(/loading/i)
        expect(loadingButton).toBeVisible()
        expect(firstRender).toMatchDiffSnapshot(asFragment())
    })

    test('should update the input field correctly, then invokes afterInputChanged() on change events', () => {
        const afterInputChanged = jest.fn()
        const value = 'test'
        const { getByRole, asFragment } = render(
            <SearchBox afterInputChanged={afterInputChanged} />
        )
        const input = getByRole('textbox')
        const firstRender = asFragment()
        fireEvent.change(input, { target: { value } })
        expect(input).toHaveValue(value)
        expect(firstRender).toMatchDiffSnapshot(asFragment())
        expect(afterInputChanged).toHaveBeenCalledWith(value)
    })

    test('should clear the input value and call afterButtonClicked on button clicks', async () => {
        const afterButtonClicked = jest.fn()
        const value = 'test'
        const { getByRole, asFragment } = render(
            <SearchBox afterButtonClicked={afterButtonClicked} />
        )
        const input = getByRole('textbox')
        const button = getByRole('button')
        fireEvent.change(input, { target: { value } })
        expect(input).toHaveValue(value)
        const firstRender = asFragment()
        fireEvent.click(button)
        expect(input).toHaveValue('')
        expect(afterButtonClicked).toHaveBeenCalledTimes(1)
        expect(firstRender).toMatchDiffSnapshot(asFragment())
    })

    test('it shoud calls afterButtonClicked on hitting enter key', async () => {
        const afterButtonClicked = jest.fn()
        const value = 'test'
        const { getByRole } = render(
            <SearchBox afterButtonClicked={afterButtonClicked} />
        )
        const input = getByRole('textbox')
        fireEvent.change(input, { target: { value } })
        expect(input).toHaveValue(value)
        fireEvent.keyDown(input, { key: EVENT_KEYS.enter })
        expect(input).toHaveValue('')
        expect(afterButtonClicked).toHaveBeenCalledTimes(1)
    })
})
