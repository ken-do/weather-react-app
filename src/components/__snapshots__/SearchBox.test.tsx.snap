// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox should clear the input value and call afterButtonClicked on button clicks 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -7,11 +7,11 @@
      >
        <input
          class=\\"border-right-0 rounded-0 form-control\\"
          placeholder=\\"Enter a location...\\"
          type=\\"text\\"
-         value=\\"test\\"
+         value=\\"\\"
        />
        <button
          class=\\"btn btn-success rounded-0\\"
          type=\\"button\\"
        >"
`;

exports[`SearchBox should enlarge when passing in a lg size prop 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,11 +1,11 @@
  <DocumentFragment>
    <form
      class=\\"flex-nowrap form-inline\\"
    >
      <div
-       class=\\"inputGroup input-group\\"
+       class=\\"inputGroup input-group input-group-lg\\"
      >
        <input
          class=\\"border-right-0 rounded-0 form-control\\"
          placeholder=\\"Enter a location...\\"
          type=\\"text\\""
`;

exports[`SearchBox should update the button lable based on isLoading value 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -13,10 +13,15 @@
        />
        <button
          class=\\"btn btn-success rounded-0\\"
          type=\\"button\\"
        >
-         Search
+         <span
+           aria-hidden=\\"true\\"
+           class=\\"mr-1 spinner-grow spinner-grow-sm\\"
+           role=\\"status\\"
+         />
+         Loading...
        </button>
      </div>
    </form>
  </DocumentFragment>"
`;

exports[`SearchBox should update the input field correctly, then invokes afterInputChanged() on change events 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -7,11 +7,11 @@
      >
        <input
          class=\\"border-right-0 rounded-0 form-control\\"
          placeholder=\\"Enter a location...\\"
          type=\\"text\\"
-         value=\\"\\"
+         value=\\"test\\"
        />
        <button
          class=\\"btn btn-success rounded-0\\"
          type=\\"button\\"
        >"
`;
